wallet/src/de/schildbach/wallet/ui/dashpay/CreateIdentityService.kt

```diff
@@ -325,11 +337,29 @@ class CreateIdentityService : LifecycleService() {
         if (blockchainIdentityData.creationState <= CreationState.UPGRADING_WALLET) {
             platformRepo.updateIdentityCreationState(blockchainIdentityData, CreationState.UPGRADING_WALLET)
             val seed = decryptSeed(backgroundHandler, wallet, encryptionKey)
-            platformRepo.addWalletAuthenticationKeysAsync(seed, encryptionKey)
+            platformRepo.addWalletKeyChainsAsync(seed, encryptionKey)
         }
 
         val blockchainIdentity = platformRepo.initBlockchainIdentity(blockchainIdentityData, wallet)
+        // look for the credit funding tx in case there was an error in the next step previously
+        for (tx in wallet.creditFundingTransactions) {
+            tx as CreditFundingTransaction
+            if (walletApplication.wallet!!.blockchainIdentityFundingKeyChain.findKeyFromPubHash(tx.creditBurnPublicKeyId.bytes) != null) {
+                blockchainIdentity.initializeCreditFundingTransaction(tx)
+            }
+        }
+
+        if (blockchainIdentityData.creationState <= CreationState.MIXING_FUNDS) {
+            platformRepo.updateIdentityCreationState(blockchainIdentityData, CreationState.MIXING_FUNDS)
+            coinJoinService.configureMixing(
+                Constants.DASH_PAY_FEE,
+                { encryptionKey },
+                { it.decrypt(encryptionKey) })
+
+            coinJoinService.prepareAndStartMixing()
 
+            coinJoinService.waitForMixing();
+        }
```

Comment:
waitForMixing() is a blocking call.  I suppose that is fine because the calling method is also a `suspend` method that is waiting on my other functions to return control before moving on to the next step.

Reply:
It could still block: `suspend` method doesn't guarantee that the code will be executed in another thread - if it's called from the main thread, it'll be executed in the main thread until the context is switched.

I see that the `serviceScope` in which a lot of suspend methods are launched is created using `Dispatchers.Main` - any reason for this? Perhaps `.Unconfined` is a better choice.

Reply:
There is no reason for using `Dispatchers.Main` unless it is the default if we don't specify another.

Reply:
I am pretty sure I don't want to use `Dispatchers.Main` because I was getting some exceptions from Room DB related calls.

Reply:
Is there a way to specify a dispatcher in this part of the code?
```
        blockchainStateProvider.observeNetworkStatus()
            .filterNot { it == NetworkStatus.UNKNOWN }
            .onEach { status ->
                updateNetworkStatus(status)
            }
            .launchIn(coroutineScope)
 ```

Reply:
@HashEngineering my point is that right now, the scope created with `.Main`:
`private var serviceScope = CoroutineScope(serviceJob + Dispatchers.Main)`

That means it's executed in the main thread. Perhaps `.IO` or `.Unconfined` which is the default is better here.

Reply:
Thanks.  This is what I needed to know to understand.

Reply:
Changing to `.IO` results in these exceptions.  This gives us an opportunity to move away from these `AsyncTask` methods such as `DecryptSeedTask` and use coroutines instead.

```
[DefaultDispatcher-worker-1] [UPGRADING_WALLET(error)]
java.lang.NullPointerException: Attempt to read from field 'android.os.MessageQueue android.os.Looper.mQueue' on a null object reference
	at android.os.Handler.<init>(Handler.java:257) ~[na:0.0]
	at android.os.Handler.<init>(Handler.java:162) ~[na:0.0]
	at de.schildbach.wallet.payments.DecryptSeedTask.<init>(DecryptSeedTask.java:39) ~[na:0.0]
	at de.schildbach.wallet.ui.dashpay.CreateIdentityService$decryptSeed$2$1.<init>(CreateIdentityService.kt:795) ~[na:0.0]
	at de.schildbach.wallet.ui.dashpay.CreateIdentityService.decryptSeed(CreateIdentityService.kt:795) ~[na:0.0]
	at de.schildbach.wallet.ui.dashpay.CreateIdentityService.createIdentity(CreateIdentityService.kt:339) ~[na:0.0]
	at de.schildbach.wallet.ui.dashpay.CreateIdentityService.access$createIdentity(CreateIdentityService.kt:58) ~[na:0.0]
	at de.schildbach.wallet.ui.dashpay.CreateIdentityService$createIdentity$1.invokeSuspend(Unknown Source:16) ~[na:0.0]
```

Reply:
Could be a good idea. `DecryptSeedTask` seems to be quite simple class whose purpose is to run `decrypt` in the background. We can just launch it in a `withContext` instead. This method that wraps it in a coroutine is also unnecessary:
```
/**
* Wraps callbacks of DecryptSeedTask as Coroutine
*/
private suspend fun decryptSeed(handler: Handler, wallet: Wallet, encryptionKey: KeyParameter): DeterministicSeed 
```

Instead, we can just run it like this:
```
withContext(Dispatchers.Default) {
   val decryptedSeed = seed.decrypt(keyCrypter, null, keyParameter)
}
```

For `DeriveKeyTask`, there is an analog function in the `SecurityFunctions` that can also be run in a context.